// @vortex77/gitkit/lib/config.js
const fs = require('fs');
const path = require('path');

class Config {
    constructor() {
        this.configFileName = '.gitkit.json';
        this.configPath = path.join(process.cwd(), this.configFileName);
    }

    configExists() {
        return fs.existsSync(this.configPath);
    }

    loadConfig() {
        try {
            if (!this.configExists()) {
                return null;
            }
            const data = fs.readFileSync(this.configPath, 'utf8');
            return JSON.parse(data);
        } catch (error) {
            console.error('Error loading config:', error.message);
            return null;
        }
    }

    saveConfig(config) {
        try {
            const data = JSON.stringify(config, null, 2);
            fs.writeFileSync(this.configPath, data, 'utf8');
            return true;
        } catch (error) {
            console.error('Error saving config:', error.message);
            return false;
        }
    }

    getConfig(key) {
        const config = this.loadConfig();
        if (!config) return null;
        return key ? config[key] : config;
    }

    updateConfig(updates) {
        const config = this.loadConfig() || {};
        const newConfig = { ...config, ...updates };
        return this.saveConfig(newConfig);
    }

    deleteConfig() {
        try {
            if (this.configExists()) {
                fs.unlinkSync(this.configPath);
                return true;
            }
            return false;
        } catch (error) {
            console.error('Error deleting config:', error.message);
            return false;
        }
    }
}

module.exports = Config;
