// @vortex77/gitkit/lib/git-operations.js
const simpleGit = require('simple-git');
const chalk = require('chalk');
const ora = require('ora');

class GitOperations {
    constructor(config = null) {
        this.git = simpleGit();
        this.config = config;
    }

    async smartPush(commitMessage) {
        const spinner = ora('Preparing to push...').start();

        try {
            // Check if there are changes
            const status = await this.git.status();

            if (status.files.length === 0) {
                spinner.warn(chalk.yellow('No changes to commit'));
                return { success: false, message: 'No changes detected' };
            }

            // Add all changes
            spinner.text = 'Adding files...';
            await this.git.add('.');

            // Commit
            spinner.text = 'Committing changes...';
            await this.git.commit(commitMessage);

            // Push
            spinner.text = 'Pushing to remote...';
            await this.git.push();

            spinner.succeed(chalk.green('✓ Successfully pushed!'));
            return { success: true, message: 'Changes pushed successfully' };

        } catch (error) {
            spinner.fail(chalk.red('✗ Push failed'));
            return { success: false, message: error.message };
        }
    }

    async listBranches() {
        try {
            const branches = await this.git.branch();
            return {
                success: true,
                current: branches.current,
                all: branches.all,
                branches: branches.branches
            };
        } catch (error) {
            return { success: false, message: error.message };
        }
    }

    async createBranch(branchName) {
        const spinner = ora(`Creating branch '${branchName}'...`).start();

        try {
            await this.git.checkoutLocalBranch(branchName);
            spinner.succeed(chalk.green(`✓ Branch '${branchName}' created and checked out`));
            return { success: true, message: `Branch ${branchName} created` };
        } catch (error) {
            spinner.fail(chalk.red(`✗ Failed to create branch '${branchName}'`));
            return { success: false, message: error.message };
        }
    }

    async switchBranch(branchName) {
        const spinner = ora(`Switching to branch '${branchName}'...`).start();

        try {
            await this.git.checkout(branchName);
            spinner.succeed(chalk.green(`✓ Switched to branch '${branchName}'`));
            return { success: true, message: `Switched to ${branchName}` };
        } catch (error) {
            spinner.fail(chalk.red(`✗ Failed to switch to '${branchName}'`));
            return { success: false, message: error.message };
        }
    }

    async deleteBranch(branchName, force = false) {
        const spinner = ora(`Deleting branch '${branchName}'...`).start();

        try {
            await this.git.deleteLocalBranch(branchName, force);
            spinner.succeed(chalk.green(`✓ Branch '${branchName}' deleted`));
            return { success: true, message: `Branch ${branchName} deleted` };
        } catch (error) {
            spinner.fail(chalk.red(`✗ Failed to delete branch '${branchName}'`));
            return { success: false, message: error.message };
        }
    }

    async mergeBranch(branchName) {
        const spinner = ora(`Merging branch '${branchName}'...`).start();

        try {
            const result = await this.git.merge([branchName]);

            if (result.conflicts.length > 0) {
                spinner.warn(chalk.yellow(`⚠ Merge conflicts detected in ${result.conflicts.length} file(s)`));
                return {
                    success: false,
                    conflicts: result.conflicts,
                    message: 'Merge conflicts need to be resolved'
                };
            }

            spinner.succeed(chalk.green(`✓ Successfully merged '${branchName}'`));
            return { success: true, message: `Merged ${branchName}` };
        } catch (error) {
            spinner.fail(chalk.red(`✗ Failed to merge '${branchName}'`));
            return { success: false, message: error.message };
        }
    }

    async getStatus() {
        try {
            const status = await this.git.status();
            return {
                success: true,
                status: status
            };
        } catch (error) {
            return { success: false, message: error.message };
        }
    }

    async displayStatus() {
        const spinner = ora('Fetching status...').start();

        try {
            const status = await this.git.status();
            spinner.stop();

            console.log('\n' + chalk.bold.cyan('=== Git Status ===\n'));
            console.log(chalk.bold('Branch: ') + chalk.green(status.current));
            console.log(chalk.bold('Tracking: ') + (status.tracking || chalk.gray('Not tracking')));

            if (status.ahead > 0) {
                console.log(chalk.yellow(`↑ ${status.ahead} commit(s) ahead`));
            }
            if (status.behind > 0) {
                console.log(chalk.yellow(`↓ ${status.behind} commit(s) behind`));
            }

            if (status.files.length > 0) {
                console.log('\n' + chalk.bold('Changes:'));
                status.files.forEach(file => {
                    const symbol = this._getFileStatusSymbol(file);
                    console.log(`  ${symbol} ${file.path}`);
                });
            } else {
                console.log('\n' + chalk.green('✓ Working tree clean'));
            }

            console.log('');
            return { success: true };
        } catch (error) {
            spinner.fail(chalk.red('✗ Failed to get status'));
            return { success: false, message: error.message };
        }
    }

    _getFileStatusSymbol(file) {
        if (file.working_dir === 'M') return chalk.yellow('M');
        if (file.working_dir === 'D') return chalk.red('D');
        if (file.working_dir === '?') return chalk.gray('?');
        if (file.index === 'A') return chalk.green('A');
        if (file.index === 'M') return chalk.green('M');
        return chalk.gray('·');
    }

    // Basic Git Commands
    async pull(remote = 'origin', branch = null) {
        const spinner = ora('Pulling changes...').start();

        try {
            const branchName = branch || (this.config && this.config.defaultBranch) || 'main';
            await this.git.pull(remote, branchName);
            spinner.succeed(chalk.green(`✓ Pulled from ${remote}/${branchName}`));
            return { success: true, message: 'Pull successful' };
        } catch (error) {
            spinner.fail(chalk.red('✗ Pull failed'));
            return { success: false, message: error.message };
        }
    }

    async clone(repoUrl, targetDir = null) {
        const spinner = ora(`Cloning ${repoUrl}...`).start();

        try {
            await this.git.clone(repoUrl, targetDir);
            spinner.succeed(chalk.green('✓ Repository cloned successfully'));
            return { success: true, message: 'Clone successful' };
        } catch (error) {
            spinner.fail(chalk.red('✗ Clone failed'));
            return { success: false, message: error.message };
        }
    }

    async init() {
        const spinner = ora('Initializing Git repository...').start();

        try {
            await this.git.init();
            spinner.succeed(chalk.green('✓ Git repository initialized'));
            return { success: true, message: 'Init successful' };
        } catch (error) {
            spinner.fail(chalk.red('✗ Init failed'));
            return { success: false, message: error.message };
        }
    }

    async log(count = 10) {
        const spinner = ora('Fetching commit history...').start();

        try {
            const logs = await this.git.log({ maxCount: count });
            spinner.stop();

            console.log('\n' + chalk.bold.cyan('=== Commit History ===\n'));

            if (logs.all.length === 0) {
                console.log(chalk.gray('No commits yet\n'));
                return { success: true, logs: [] };
            }

            logs.all.forEach((commit, index) => {
                console.log(chalk.yellow(`${commit.hash.substring(0, 7)}`) + chalk.gray(' - ') + commit.message);
                console.log(chalk.gray(`  ${commit.author_name} <${commit.author_email}>`));
                console.log(chalk.gray(`  ${new Date(commit.date).toLocaleString()}\n`));
            });

            return { success: true, logs: logs.all };
        } catch (error) {
            spinner.fail(chalk.red('✗ Failed to fetch logs'));
            return { success: false, message: error.message };
        }
    }

    async diff(file = null) {
        const spinner = ora('Fetching diff...').start();

        try {
            const diffSummary = file ? await this.git.diff([file]) : await this.git.diff();
            spinner.stop();

            if (!diffSummary) {
                console.log(chalk.gray('\nNo changes\n'));
                return { success: true, diff: '' };
            }

            console.log('\n' + chalk.bold.cyan('=== Changes ===\n'));
            console.log(diffSummary);
            console.log('');

            return { success: true, diff: diffSummary };
        } catch (error) {
            spinner.fail(chalk.red('✗ Failed to get diff'));
            return { success: false, message: error.message };
        }
    }

    async fetch(remote = 'origin') {
        const spinner = ora(`Fetching from ${remote}...`).start();

        try {
            await this.git.fetch(remote);
            spinner.succeed(chalk.green(`✓ Fetched from ${remote}`));
            return { success: true, message: 'Fetch successful' };
        } catch (error) {
            spinner.fail(chalk.red('✗ Fetch failed'));
            return { success: false, message: error.message };
        }
    }
}

module.exports = GitOperations;
