// @vortex77/gitkit/lib/prompts.js
const inquirer = require('inquirer');
const chalk = require('chalk');

const prompts = {
    mainMenu: async () => {
        return inquirer.prompt([
            {
                type: 'list',
                name: 'action',
                message: chalk.green('[') + chalk.white('SELECT OPERATION') + chalk.green(']'),
                choices: [
                    { name: chalk.green('[1]') + chalk.white(' Smart Push') + chalk.gray(' → add, commit, push'), value: 'push' },
                    { name: chalk.green('[2]') + chalk.white(' Branch Management') + chalk.gray(' → create, switch, delete'), value: 'branch' },
                    { name: chalk.green('[3]') + chalk.white(' Merge Branch') + chalk.gray(' → merge operations'), value: 'merge' },
                    { name: chalk.green('[4]') + chalk.white(' Repository Status') + chalk.gray(' → view changes'), value: 'status' },
                    { name: chalk.red('[X]') + chalk.white(' Exit'), value: 'exit' }
                ]
            }
        ]);
    },

    commitMessage: async () => {
        return inquirer.prompt([
            {
                type: 'input',
                name: 'message',
                message: chalk.green('[') + chalk.white('COMMIT MESSAGE') + chalk.green(']'),
                validate: (input) => {
                    if (input.trim().length === 0) {
                        return chalk.red('[!] Commit message cannot be empty');
                    }
                    return true;
                }
            }
        ]);
    },

    branchMenu: async () => {
        return inquirer.prompt([
            {
                type: 'list',
                name: 'action',
                message: chalk.green('[') + chalk.white('BRANCH OPERATIONS') + chalk.green(']'),
                choices: [
                    { name: chalk.green('[L]') + chalk.white(' List branches') + chalk.gray(' → show all'), value: 'list' },
                    { name: chalk.green('[C]') + chalk.white(' Create new branch') + chalk.gray(' → checkout new'), value: 'create' },
                    { name: chalk.green('[S]') + chalk.white(' Switch branch') + chalk.gray(' → checkout existing'), value: 'switch' },
                    { name: chalk.red('[D]') + chalk.white(' Delete branch') + chalk.gray(' → remove'), value: 'delete' },
                    { name: chalk.gray('[←]') + chalk.white(' Back'), value: 'back' }
                ]
            }
        ]);
    },

    branchName: async () => {
        return inquirer.prompt([
            {
                type: 'input',
                name: 'name',
                message: chalk.green('[') + chalk.white('BRANCH NAME') + chalk.green(']'),
                validate: (input) => {
                    if (input.trim().length === 0) {
                        return chalk.red('[!] Branch name cannot be empty');
                    }
                    if (!/^[a-zA-Z0-9/_-]+$/.test(input)) {
                        return chalk.red('[!] Invalid characters (use: a-z, 0-9, /, -, _)');
                    }
                    return true;
                }
            }
        ]);
    },

    selectBranch: async (branches) => {
        return inquirer.prompt([
            {
                type: 'list',
                name: 'branch',
                message: chalk.green('[') + chalk.white('SELECT BRANCH') + chalk.green(']'),
                choices: branches.map(b => ({
                    name: b.current ? chalk.green('→ ') + chalk.white(b.name) : chalk.gray('  ') + chalk.white(b.name),
                    value: b.name
                }))
            }
        ]);
    },

    confirm: async (message) => {
        return inquirer.prompt([
            {
                type: 'confirm',
                name: 'confirmed',
                message: chalk.yellow('[?] ') + chalk.white(message),
                default: false
            }
        ]);
    },

    forceDelete: async () => {
        return inquirer.prompt([
            {
                type: 'confirm',
                name: 'force',
                message: chalk.red('[!] ') + chalk.white('Branch has unmerged changes. Force delete?'),
                default: false
            }
        ]);
    }
};

module.exports = prompts;
