// @vortex77/gitkit/lib/setup.js
const inquirer = require('inquirer');
const chalk = require('chalk');
const ora = require('ora');
const Config = require('./config');

class Setup {
    constructor() {
        this.config = new Config();
    }

    displayWelcome() {
        console.clear();
        console.log(chalk.green(`
  ╔═══════════════════════════════════════════════╗
  ║                                               ║
  ║     GITKIT CONFIGURATION WIZARD v1.0.0        ║
  ║                                               ║
  ╚═══════════════════════════════════════════════╝
        `));
        console.log(chalk.gray('  [') + chalk.green('*') + chalk.gray('] Initializing setup sequence...'));
        console.log(chalk.gray('  [') + chalk.green('+') + chalk.gray('] Configuration mode: INTERACTIVE\n'));
    }

    async run() {
        this.displayWelcome();

        const answers = await inquirer.prompt([
            {
                type: 'input',
                name: 'repoUrl',
                message: chalk.green('[') + chalk.white('REPO URL') + chalk.green(']') + chalk.gray(' (optional)'),
                default: '',
                validate: (input) => {
                    if (!input) return true;
                    if (input.includes('github.com') || input.includes('gitlab.com') ||
                        input.includes('bitbucket.org') || input.startsWith('git@') ||
                        input.startsWith('https://') || input.startsWith('http://')) {
                        return true;
                    }
                    return chalk.red('[!] Invalid Git repository URL');
                }
            },
            {
                type: 'input',
                name: 'defaultBranch',
                message: chalk.green('[') + chalk.white('DEFAULT BRANCH') + chalk.green(']'),
                default: 'main'
            },
            {
                type: 'confirm',
                name: 'autoPush',
                message: chalk.green('[') + chalk.white('AUTO-PUSH') + chalk.green(']') + chalk.gray(' after commit?'),
                default: true
            },
            {
                type: 'confirm',
                name: 'autoFetch',
                message: chalk.green('[') + chalk.white('AUTO-FETCH') + chalk.green(']') + chalk.gray(' before operations?'),
                default: false
            },
            {
                type: 'list',
                name: 'commitStyle',
                message: chalk.green('[') + chalk.white('COMMIT STYLE') + chalk.green(']'),
                choices: [
                    { name: chalk.white('Conventional') + chalk.gray(' (feat:, fix:, etc.)'), value: 'conventional' },
                    { name: chalk.white('Simple') + chalk.gray(' (free-form)'), value: 'simple' },
                    { name: chalk.white('Custom prefix'), value: 'custom' }
                ],
                default: 'simple'
            }
        ]);

        if (answers.commitStyle === 'custom') {
            const { customPrefix } = await inquirer.prompt([
                {
                    type: 'input',
                    name: 'customPrefix',
                    message: chalk.green('[') + chalk.white('CUSTOM PREFIX') + chalk.green(']'),
                    default: '[PROJECT]'
                }
            ]);
            answers.customPrefix = customPrefix;
        }

        const spinner = ora(chalk.gray('[*] Saving configuration...')).start();

        const config = {
            repoUrl: answers.repoUrl || null,
            defaultBranch: answers.defaultBranch,
            autoPush: answers.autoPush,
            autoFetch: answers.autoFetch,
            commitStyle: answers.commitStyle,
            customPrefix: answers.customPrefix || null,
            setupDate: new Date().toISOString()
        };

        const saved = this.config.saveConfig(config);

        if (saved) {
            spinner.succeed(chalk.green('[+] Configuration saved successfully'));
            console.log('\n' + chalk.gray('  [') + chalk.green('*') + chalk.gray('] Setup complete!'));
            console.log(chalk.gray('  [') + chalk.white('i') + chalk.gray('] Reconfigure anytime: ') + chalk.white('gitkit setup\n'));
            return config;
        } else {
            spinner.fail(chalk.red('[!] Failed to save configuration'));
            return null;
        }
    }

    async reconfigure() {
        console.log(chalk.yellow('\n  [!] This will overwrite your existing configuration.\n'));
        const { confirm } = await inquirer.prompt([
            {
                type: 'confirm',
                name: 'confirm',
                message: chalk.yellow('[?] ') + chalk.white('Continue?'),
                default: false
            }
        ]);

        if (!confirm) {
            console.log(chalk.gray('\n  [-] Cancelled.\n'));
            return null;
        }

        return this.run();
    }
}

module.exports = Setup;
