// @vortex77/gitkit/lib/ui.js
const chalk = require('chalk');
const GitOperations = require('./git-operations');
const prompts = require('./prompts');

class UI {
    constructor() {
        this.git = new GitOperations();
    }

    displayWelcome() {
        console.clear();
        console.log(chalk.green(`
  ██████╗ ██╗████████╗██╗  ██╗██╗████████╗
 ██╔════╝ ██║╚══██╔══╝██║ ██╔╝██║╚══██╔══╝
 ██║  ███╗██║   ██║   █████╔╝ ██║   ██║   
 ██║   ██║██║   ██║   ██╔═██╗ ██║   ██║   
 ╚██████╔╝██║   ██║   ██║  ██╗██║   ██║   
  ╚═════╝ ╚═╝   ╚═╝   ╚═╝  ╚═╝╚═╝   ╚═╝   
        `));
        console.log(chalk.gray('  ═══════════════════════════════════════════════'));
        console.log(chalk.gray('  │') + chalk.white('  Advanced Git Toolkit v1.0.0') + chalk.gray('               │'));
        console.log(chalk.gray('  │') + chalk.green('  [+]') + chalk.white(' System loaded') + chalk.gray('                         │'));
        console.log(chalk.gray('  ═══════════════════════════════════════════════\n'));
    }

    async start() {
        this.displayWelcome();
        await this.mainLoop();
    }

    async mainLoop() {
        while (true) {
            console.log(chalk.gray('  ═══════════════════════════════════════════════'));
            const { action } = await prompts.mainMenu();
            console.log(chalk.gray('  ═══════════════════════════════════════════════\n'));

            switch (action) {
                case 'push':
                    await this.handleSmartPush();
                    break;
                case 'branch':
                    await this.handleBranchManagement();
                    break;
                case 'merge':
                    await this.handleMerge();
                    break;
                case 'status':
                    await this.handleStatus();
                    break;
                case 'exit':
                    console.log(chalk.gray('\n  ═══════════════════════════════════════════════'));
                    console.log(chalk.gray('  [') + chalk.red('!') + chalk.gray('] Shutting down GitKit...'));
                    console.log(chalk.gray('  ═══════════════════════════════════════════════\n'));
                    process.exit(0);
            }
        }
    }

    async handleSmartPush() {
        const { message } = await prompts.commitMessage();
        const result = await this.git.smartPush(message);

        if (!result.success) {
            console.log(chalk.red(`\n  [!] ${result.message}\n`));
        }
    }

    async handleBranchManagement() {
        while (true) {
            const { action } = await prompts.branchMenu();

            if (action === 'back') break;

            switch (action) {
                case 'list':
                    await this.listBranches();
                    break;
                case 'create':
                    await this.createBranch();
                    break;
                case 'switch':
                    await this.switchBranch();
                    break;
                case 'delete':
                    await this.deleteBranch();
                    break;
            }
        }
    }

    async listBranches() {
        const result = await this.git.listBranches();

        if (result.success) {
            console.log('\n' + chalk.gray('  ═══════════════════════════════════════════════'));
            console.log(chalk.green('  [*] BRANCHES'));
            console.log(chalk.gray('  ───────────────────────────────────────────────'));
            Object.entries(result.branches).forEach(([name, info]) => {
                const prefix = name === result.current ? chalk.green('  →  ') : chalk.gray('     ');
                console.log(prefix + chalk.white(name));
            });
            console.log(chalk.gray('  ═══════════════════════════════════════════════\n'));
        } else {
            console.log(chalk.red(`\n  [!] ERROR: ${result.message}\n`));
        }
    }

    async createBranch() {
        const { name } = await prompts.branchName();
        await this.git.createBranch(name);
    }

    async switchBranch() {
        const branchesResult = await this.git.listBranches();

        if (!branchesResult.success) {
            console.log(chalk.red(`\n  [!] ${branchesResult.message}\n`));
            return;
        }

        const branchList = Object.entries(branchesResult.branches).map(([name, info]) => ({
            name: name,
            current: name === branchesResult.current
        }));

        const { branch } = await prompts.selectBranch(branchList);

        if (branch === branchesResult.current) {
            console.log(chalk.yellow('\n  [!] Already on this branch\n'));
            return;
        }

        await this.git.switchBranch(branch);
    }

    async deleteBranch() {
        const branchesResult = await this.git.listBranches();

        if (!branchesResult.success) {
            console.log(chalk.red(`\n  [!] ${branchesResult.message}\n`));
            return;
        }

        const branchList = Object.entries(branchesResult.branches)
            .filter(([name]) => name !== branchesResult.current)
            .map(([name, info]) => ({
                name: name,
                current: false
            }));

        if (branchList.length === 0) {
            console.log(chalk.yellow('\n  [!] No other branches to delete\n'));
            return;
        }

        const { branch } = await prompts.selectBranch(branchList);
        const { confirmed } = await prompts.confirm(`Delete branch '${branch}'?`);

        if (!confirmed) {
            console.log(chalk.gray('\n  [-] Cancelled\n'));
            return;
        }

        const result = await this.git.deleteBranch(branch);

        if (!result.success && result.message.includes('not fully merged')) {
            const { force } = await prompts.forceDelete();
            if (force) {
                await this.git.deleteBranch(branch, true);
            }
        }
    }

    async handleMerge() {
        const branchesResult = await this.git.listBranches();

        if (!branchesResult.success) {
            console.log(chalk.red(`\n  [!] ERROR: ${branchesResult.message}\n`));
            return;
        }

        const branchList = Object.entries(branchesResult.branches)
            .filter(([name]) => name !== branchesResult.current)
            .map(([name, info]) => ({
                name: name,
                current: false
            }));

        if (branchList.length === 0) {
            console.log(chalk.yellow('\n  [!] No other branches available to merge\n'));
            return;
        }

        const { branch } = await prompts.selectBranch(branchList);
        const result = await this.git.mergeBranch(branch);

        if (!result.success && result.conflicts) {
            console.log('\n' + chalk.gray('  ═══════════════════════════════════════════════'));
            console.log(chalk.red('  [!] MERGE CONFLICTS DETECTED'));
            console.log(chalk.gray('  ───────────────────────────────────────────────'));
            result.conflicts.forEach(file => {
                console.log(chalk.yellow(`      → ${file}`));
            });
            console.log(chalk.gray('  ───────────────────────────────────────────────'));
            console.log(chalk.gray('  [*] Action required: Resolve conflicts manually'));
            console.log(chalk.gray('  ═══════════════════════════════════════════════\n'));
        }
    }

    async handleStatus() {
        await this.git.displayStatus();
    }
}

module.exports = UI;
